shelley-ma/formal-spec/value.tex

```diff
@@ -1,77 +1,239 @@
+  single term)
+\end{itemize}
+
+Below we give the definitions of all the functions that must be defined on
+$\Coin$ and $\Value$ in order for them to have the structure of $\ValMonoid$.
+In Section \ref{sec:other-valmonoids} we give several other types which we can meaningfully
+support the definition of the required functions (addition, size, etc.), including
+an optimized representation that more accurately represents the ledger implementation
+of the multi-asset type.
+These types are convertible to and from the $\Value$ type, and appear in other
+parts of the system as a multi-asset representation which is more suitable in particular
+usecases.
+
+\subsection{$\ValMonoid$}
+
+Figure \ref{fig:ValMonoid} gives names and types of functions which
+make up the $\ValMonoid$ structure.
+
+%%
+%% Figure ValMonoid and its Functions
+%%
+\begin{figure}[htb]
+  \begin{align*}
+      & \mathsf{ValType} ~\in~ \mathsf{Type}  \\
+      & \text{A type}
+      \nextdef
+      %
+      & \fun{zero} ~\in~ \mathsf{ValType} \\
+      & \text{Additive identity of $\mathsf{ValType}$}
+      \nextdef
+      %
+      & (+) ~\in~ \mathsf{ValType} \to \mathsf{ValType} \to \mathsf{ValType}\\
+      & \text{Addition on $\mathsf{ValType}$}
+      \nextdef
+      %
+      & (*) ~\in~ \mathsf{Integer} \to \mathsf{ValType} \to \mathsf{ValType}\\
+      & \text{Scale $\mathsf{ValType}$ value by integral constant}
+      \nextdef
+      %
+      & (-) ~\in~ \mathsf{ValType} \to \mathsf{ValType} \to \mathsf{ValType}\\
+      & v_1~-~v_2 ~=~ v_1~+~((-1)~*~v_2)\\
+      & \text{Subtraction on $\mathsf{ValType}$}
+      \nextdef
+      %
+      & \fun{isZero} ~\in~ \mathsf{ValType} \to \Bool\\
+      & \fun{isZero}~v~=~(v~==~\fun{zero}) \\
+      & \text{Is a given $\mathsf{ValType}$ term the additive identity?}
+      \nextdef
+      %
+      & \fun{coin} ~\in~ \mathsf{ValType} \to \Coin\\
+      & \text{Return the Ada contained in a $\mathsf{ValType}$ term}
+      \nextdef
+      %
+      & \fun{inject} ~\in~ \Coin \to \mathsf{ValType}\\
+      & \text{Create a $\mathsf{ValType}$ term containing only this amount of Ada}
+      \nextdef
+      %
+      & \fun{adaOnly} ~\in~ \mathsf{ValType} \to \Bool \\
+      & \fun{adaOnly} =
+        \begin{cases}
+          \True & \fun{inject} \circ \fun{coin}~ = ~\fun{id}_{\mathsf{ValType}} \\
+          \False & \text{otherwise}
+        \end{cases} \\
+      & \text{Check if a given $\mathsf{ValType}$ term contains only Ada}
+      \nextdef
+      %
+      & \fun{modifyCoin} ~\in~ (\Coin \to \Coin) \to (\mathsf{ValType} \to \mathsf{ValType})\\
+      & \text{Modify the Ada part of $\mathsf{ValType}$}
+      \nextdef
+      %
+      & \fun{size} ~\in~ \mathsf{ValType} \to \MemoryEstimate\\
+      & \text{Return the size in words of a $\mathsf{ValType}$ term}
+      \nextdef
+      %
+      & \fun{pointwise} ~\in~ (\Integer \to \Integer \to \Bool) \to (\mathsf{ValType} \to \mathsf{ValType} \to \Bool)\\
```

Comment:
Also unused

Reply:
it is technically used where we use "<=" on Values. I am not sure if I should make a note of that or replace the math symbol with `pointwise`

Reply:
I think we want to put the requirement of a partial order on the monoid aspect. I.e. instead of saying we want a monoid with additional functions, we want a partially ordered monoid with additional functions. In fact, a monoid with the module structure given by `*` is already a group, and we probably want it to be commutative. So, I guess we want to say `partially ordered abelian group with extra stuff`. Then we can also drop a huge bunch of stuff in this figure (`zero, +, *, -, <=`)

Reply:
I dont think we should drop any of this, since we do actually need to specify it, and it might be better not to assume that people know these things are implied by partially ordered abelian group structure

Reply:
Instead of putting it in the figure, there could be a sentence or short paragraph reminding the reader what a partially ordered abelian group is

Reply:
So the thing with `pointwise` is that it's used to define the relations of a specific `ValMonoid` (i.e. `Value`). `ValMonoid` should not require `pointwise` to be defined, but it should require `<=` etc. to be defined.

Reply:
i do not agree with this - i think using pointwise makes the spec more clear and easier to compare to the code

Reply:
It's a fault of the code (inherited from Haskell) though, that it isn't using things like `<=` to compare. If I'd implement the spec in another language, there's a very high likelihood that it would be using `<=` and not `pointwise (<=)`. So the code should maybe mention in a comment why it isn't using `<=` to explain this deviation.